Core/src/Geometry/Portal.cpp

```diff
@@ -0,0 +1,96 @@
+Acts::Experimental::Portal::getSharedPtr() const {
+  return shared_from_this();
+}
+
+void Acts::Experimental::Portal::assignGeometryId(
+    const GeometryIdentifier& geometryId) {
+  m_surface->assignGeometryId(geometryId);
+}
+
+void Acts::Experimental::Portal::fuse(std::shared_ptr<Portal>& other) {
+  // Determine this directioon
+  NavigationDirection tDir =
+      (m_volumeUpdators[indexFromDirection(NavigationDirection::Backward)]
+           .implementation == nullptr)
+          ? NavigationDirection::Forward
+          : NavigationDirection::Backward;
+
+  if (m_volumeUpdators[indexFromDirection(tDir)].implementation == nullptr) {
+    throw std::invalid_argument(
+        "Portal: trying to fuse portal (keep) with no links.");
+  }
+  // And now check other direction
+  NavigationDirection oDir = invertDirection(tDir);
+  if (other->m_volumeUpdators[indexFromDirection(oDir)].implementation ==
+      nullptr) {
+    throw std::runtime_error(
+        "Portal: trying to fuse portal (waste) with no links.");
+  }
+
+  auto odx = indexFromDirection(oDir);
+  m_volumeUpdators[odx] = std::move(other->m_volumeUpdators[odx]);
+  m_attachedVolumes[odx] = other->m_attachedVolumes[odx];
+
+  // And finally overwrite the original portal
+  other = getSharedPtr();
+}
+
+void Acts::Experimental::Portal::assignDetectorVolumeUpdator(
+    NavigationDirection nDir, ManagedDetectorVolumeUpdator&& dVolumeUpdator,
+    const std::vector<std::shared_ptr<DetectorVolume>>& attachedVolumes) {
+  auto idx = indexFromDirection(nDir);
+  m_volumeUpdators[idx] = std::move(dVolumeUpdator);
+  m_attachedVolumes[idx] = attachedVolumes;
+}
+
+void Acts::Experimental::Portal::assignDetectorVolumeUpdator(
+    ManagedDetectorVolumeUpdator&& dVolumeUpdator,
+    const std::vector<std::shared_ptr<DetectorVolume>>& attachedVolumes) {
+  // Check and throw exceptions
+  if (m_volumeUpdators[0u].implementation == nullptr and
+      m_volumeUpdators[1u].implementation == nullptr) {
+    throw std::runtime_error("Portal: portal has no link on either side.");
+  }
+  if (m_volumeUpdators[0u].implementation != nullptr and
+      m_volumeUpdators[1u].implementation != nullptr) {
+    throw std::runtime_error("Portal: portal already links on both sides.");
+  }
+  size_t idx = (m_volumeUpdators[0u].implementation == nullptr) ? 0u : 1u;
+  m_volumeUpdators[idx] = std::move(dVolumeUpdator);
+  m_attachedVolumes[idx] = attachedVolumes;
+}
+
+void Acts::Experimental::Portal::updateDetectorVolume(
+    const GeometryContext& gctx, NavigationState& nState) const {
+  const auto& position = nState.position;
+  const auto& direction = nState.direction;
+  const Vector3 normal = surface().normal(gctx, position);
+  NavigationDirection nDir = directionFromStepSize(normal.dot(direction));
+  const auto& vUpdator = m_volumeUpdators[indexFromDirection(nDir)].delegate;
+  if (vUpdator.connected()) {
+    vUpdator(gctx, nState);
+  } else {
+    nState.currentVolume = nullptr;
+  }
```

Comment:
Could we not by default set the updator to a free function that sets `nState.currentVolume`? Then you could remove this extra check.

Reply:
Good idea!

Reply:
I will put that on top of your new `ManagedDelegate` proposal.